### ISSUE TITLE
Syntax highlighting for code blocks

### ISSUE THREAD

**Issue by xsebek** [swarm-game/swarm#574]
- part of #545 

As a quick improvement, we should be able to parse markdown code blocks and highlight those in the game UI.

If the first and last line of `Text` is ```, the lines in the middle should be (parsed as swarm code and) highlighted.

Before:
```yaml
  - 'Example:'
  - '  move; turn left; move; turn right'
```

After:
```yaml
  - 'Example:'
  - |
    ```
    move; turn left; move; turn right
    ```
```

As a bonus, we could easily check that the code parses in the CI. #309

**Comment by byorgey** [comment]
As I mentioned in #545, I do not actually think highlighting code blocks will be easy, although I would be very happy to be proven wrong.  But I agree we should do this and at least get #309 .

**Comment by xsebek** [comment]
Well for start I would not semantically highlight the code - just make it bold or italic or something like that - simple attribute of the `txt` block. :slightly_smiling_face: And of course drop the ```.

Highlighting code semantically would require highlighting individual words like in #545, so that is harder even without the extra work of matching parsed elements to code spans and highlighting those.

**Comment by byorgey** [comment]
Ah, that makes sense, yes, that should be possible.

**Comment by xsebek** [comment]
Turns out there is a library for this:
- https://hackage.haskell.org/package/brick-skylighting

**Comment by xsebek** [comment]
Well, it also needs to work with our pretty printing. :thinking: 

The way to do that seems to be to:
1. [`annotate`](https://hackage.haskell.org/package/prettyprinter-1.7.1/docs/Prettyprinter.html#v:annotate) the `Doc`.
2. produce a [`SimpleDocStream`](https://hackage.haskell.org/package/prettyprinter-1.7.1/docs/Prettyprinter.html#t:SimpleDocStream) (we can specify line length)
3. turn that into a list (lines) of lists of text segments each with an assigned [`TokenType`](https://hackage.haskell.org/package/skylighting-core-0.13/docs/Skylighting-Types.html#t:TokenType).

**Comment by xsebek** [comment]
Since I think the implementation is clear now, I will do semantic highlighting too.

Notes:
1. start by splitting the descriptions into plaintext paragraphs and code blocks
2. parse the code blocks
  - if it does not keep the original text so we can print it plainly (e.g. while developing)
3. pretty print the code blocks (consider line length)
4. set up an integration test to check that all code blocks are valid
5. follow the steps in the comment above to highlight the output
6. profit

**Comment by xsebek** [comment]
I looked into this and ended up quite confused. 😕 (See WIP in 98ee1095beac0780ef8714601f637e98ca8688f7)

Adding annotations is simple enough, but there are more steps to the `Doc TokenType -> Widget Name` conversion:
* [x] we already produce `SimpleDocStream`
* [ ] but I have no idea how to easily turn it to `[[(TokenType, Text)]]`
* [ ] for the brick library to highlight our code, we would need some `Style` definition, which [`attrMappingsForStyle `](https://hackage.haskell.org/package/brick-skylighting-1.0/docs/Brick-Widgets-Skylighting.html#v:attrMappingsForStyle) would turn to brick attribute tree that we would add to our current one
  * [ ] given style needs to be loaded from XML with [`loadSyntaxFromFile `](https://hackage.haskell.org/package/skylighting-core-0.13.4.1/docs/Skylighting-Loader.html#v:loadSyntaxFromFile)
  * [ ] this seems like it should be configurable by users, but with default provided in our data

**Comment by xsebek** [comment]
@byorgey if you have some idea how to approach this, I would be glad to hear it. 😅 

The good news is that once we figure this out, we should get nicely coloured code widgets everywhere.
Also the Parse and Validate parts will be done as part of #1106.

The bad news is that highlighting is more work than I expected and I will not implement it in #1106. I will probably have to render inline code simply and blocks can get properly highlighted later. @byorgey so I will probably rename this issue back to "Highlight code blocks". 😅

**Comment by byorgey** [comment]
> The bad news is that highlighting is more work than I expected and I will not implement it in #1106. I will probably have to render inline code simply and blocks can get properly highlighted later. @byorgey so I will probably rename this issue back to "Highlight code blocks". 😅

Makes sense.  I don't have any particular ideas off the top of my head.  If you wanted to schedule a meeting sometime to think about it together I'd be happy to do that.

**Comment by byorgey** [comment]
> * given style needs to be loaded from XML with [`loadSyntaxFromFile `](https://hackage.haskell.org/package/skylighting-core-0.13.4.1/docs/Skylighting-Loader.html#v:loadSyntaxFromFile)

@xsebek I was looking into this and I don't think this is true: `loadSyntaxFromFile` is for loading a `Syntax`, which is ... some more complicated thing.  A `Style` is much simpler: https://hackage.haskell.org/package/skylighting-core-0.13/docs/Skylighting-Types.html#t:Style  We should be able to represent/construct one in some relatively straightforward way that allows for user customization, probably just via `ToJSON`/`FromJSON` and yaml.

Converting from `SimpleDocStream` to `[[(TokenType, Text)]]` looks straightforward.  Keep track of a stack of `TokenType`, and recurse through the `SimpleDocStream`.
- Every time you see an `SLine`, start a new sub-list.
- Emit some `Text` every time you see an `SChar` or `SText`, with the `TokenType` on top of the stack.
- Every time you see an `SAnnPush` or `SAnnPop`, push or pop from the `TokenType` stack as appropriate.

@xsebek have I missed something?
